### DISCUSSION TITLE
WithFactory configuration and open generics

### DISCUSSION THREAD

**Discussion by michaelplavnik** [z4kn4fein/stashbox#138] [General]
I was trying to use WithFactory configuration to implement delegation resolution of some singleton services to external DI system (Microsoft). It all works well for closed types, but I do not see how to make it work for open generics. Neither IDependencyResolver, or IRequestContext carry information about dependency type.

May be it is a possible to provide configuration method WithServiceProvider(serviceProvider) ?

Implementing custom resolver along the following line might be excessive in this case.

    public ServiceContext GetExpression(IResolutionStrategy resolutionStrategy, TypeInformation typeInfo, ResolutionContext resolutionContext)
    {
        var method = typeof(IServiceProvider).GetMethod("GetService")!;
        var expression = Expression.Call(Expression.Constant(_serviceProvider), method, Expression.Constant(typeInfo.Type));
        return expression.AsServiceContext();
    }

**Comment by z4kn4fein** [comment]
Hi @michaelplavnik, thanks for reaching out! 

Just to verify that I understand your case correctly, do you need an option that sets a `IServiceProvider` as a fallback for the whole container?
So, if a service is missing, the container would try to resolve it from the given `IServiceProvider`, right?

**Comment by schuettecarsten** [comment]
Thats an interesting approach because it can make it unneccessary to move all the registrations from Microsoft's ServiceCollection into Stashbox. Simply ask Microsoft for unknown service instances.

**Comment by z4kn4fein** [comment]
@schuettecarsten 
> Thats an interesting approach because it can make it unneccessary to move all the registrations from Microsoft's ServiceCollection into Stashbox. Simply ask Microsoft for unknown service instances.

If I understand the case correctly, only singletons should come from the parent `IServiceProvider`. Stashbox handles everything else, so those registrations should be copied. However, if singletons would be completely skipped from the `StashboxContainer`, an unknown type resolution delegate could resolve them from the `IServiceProvider`. Something like this:

```cs
new StashboxContainer(c => c.WithUnknownTypeResolution(opts =>
    {
        opts.WithFactory(() => serviceProvider.GetRequiredService(opts.ServiceType));
    }));
```

@michaelplavnik What do you think? It's not possible yet, but I'm already working on allowing this behavior.
You'd also be able to simplify the service descriptor copying by just filtering out singletons and populating the remaining with `container.RegisterServiceDescriptors(serviceCollection)`.

**Reply by michaelplavnik** [reply]
@z4kn4fein You are correct. The issue is only with singletons, as singletons need to be shared with part of the application that is not using Stashbox. I was able to implement everything to my liking with pure Factory configuration, except for open generics. Factory callback is called without access to actual dependency type, so it is not possible to correctly delegate to the service provider.

@schuettecarsten I have only recently discovered this gem called Stashbox, so some details in my explanation can be misleading. My understanding is that copying registrations (might be in transformed way) allows Stashbox to have complete dependency graph, thus,  validation and resolution optimizations are possible then.

**Reply by z4kn4fein** [reply]
@michaelplavnik In addition to the unknown type resolution solution, I've added the option to access the currently resolving type in a factory delegate via a `TypeInformation` parameter:
```cs
opts.WithFactory<TypeInformation>(typeInfo => serviceProvider.GetRequiredService(typeInfo.Type))
```
It's in Stashbox version `v5.10.2` and extensions `v5.2.2`.

Let me know what you think!
